substrate/frame/support/src/traits/try_runtime/mod.rs

```diff
@@ -145,29 +145,90 @@ impl core::str::FromStr for UpgradeCheckSelect {
 /// the storage items of your pallet.
 ///
 /// This hook should not alter any storage.
-pub trait TryState<BlockNumber> {
-	/// Execute the state checks.
-	fn try_state(_: BlockNumber, _: Select) -> Result<(), TryRuntimeError>;
+pub trait TryStateLogic<BlockNumber> {
+	fn try_state(_: BlockNumber) -> Result<(), TryRuntimeError>;
 }
 
 #[cfg_attr(all(not(feature = "tuples-96"), not(feature = "tuples-128")), impl_for_tuples(64))]
 #[cfg_attr(all(feature = "tuples-96", not(feature = "tuples-128")), impl_for_tuples(96))]
 #[cfg_attr(all(feature = "tuples-128"), impl_for_tuples(128))]
-impl<BlockNumber: Clone + sp_std::fmt::Debug + AtLeast32BitUnsigned> TryState<BlockNumber>
-	for Tuple
+impl<BlockNumber> TryStateLogic<BlockNumber> for Tuple
+where
+	BlockNumber: Clone + sp_std::fmt::Debug + AtLeast32BitUnsigned,
+{
+	fn try_state(n: BlockNumber) -> Result<(), TryRuntimeError> {
+		let mut errors = Vec::<TryRuntimeError>::new();
+
+		for_tuples!(#(
+			if let Err(err) = Tuple::try_state(n.clone()) {
+				errors.push(err);
+			}
+		)*);
+
+		Ok(())
+	}
+}
+
+/// Logic executed when `try-state` filters are provided in the `try-runtime` CLI.
+///
+/// Returning `true` for the provided ID will include this `try-state` logic in the overall tests
+/// performed.
+pub trait IdentifiableTryStateLogic<BlockNumber>: TryStateLogic<BlockNumber> {
+	fn matches_id(_id: &[u8]) -> bool;
+}
+
+#[cfg_attr(all(not(feature = "tuples-96"), not(feature = "tuples-128")), impl_for_tuples(64))]
+#[cfg_attr(all(feature = "tuples-96", not(feature = "tuples-128")), impl_for_tuples(96))]
+#[cfg_attr(all(feature = "tuples-128"), impl_for_tuples(128))]
+impl<BlockNumber> IdentifiableTryStateLogic<BlockNumber> for Tuple
+where
+	BlockNumber: Clone + sp_std::fmt::Debug + AtLeast32BitUnsigned,
+{
+	for_tuples!( where #( Tuple: TryStateLogic<BlockNumber> )* );
+	fn matches_id(id: &[u8]) -> bool {
+		for_tuples!( #( if Tuple::matches_id(id) { return true; } )* );
+		return false;
```

Comment:
That means that it returns true when one of the tuple elements returns true? But what if we only want to run that single tuple element instead of the whole tuple?  

I think i am missing some critical info about this merge request, which is why i dont get what it was done this way.  
Is it not enough to just add a `TryState` to the end of `Executive`?

Reply:
Perhaps the context in https://github.com/paritytech/polkadot-sdk/issues/235 could help you understand better?

The goal of this PR is to achieve exactly what you are suggesting here. Having a generic to run `try-state` checks at the runtime level. The problem with the current structure (as also mentioned in the linked issue), is that the implementation on tuples is coupled with the `PalletInfoAccess` trait, as you can see here: https://github.com/paritytech/polkadot-sdk/blob/4aa29a41cf731b8181f03168240e8dedb2adfa7a/substrate/frame/support/src/traits/try_runtime/mod.rs#L156.

This means that in order to use tuple elements that are not pallet, such as runtime-wide logic, the coupling to such a trait must be removed. And because of some requirements the `impl_for_tuples` has about knowing the type and size of the types involved, it was not possible to use anything else than a function that returns `true` if the specified identifier matches what was provided by the user in the `try-state` command.

So now if you have a tuple of, for instance, `(CustomLogic, AllPalletsWithSystem)`, the matching function will return `true` if `CustomLogic` matches, or if any pallet has the specified ID. It is assumed that no two components, be they pallets or else, have the same ID, as that would mean the latter of them would never be called.

Reply:
And this goes beyond the `AllPalletsWithSystem` usage. If even just `CustomLogic` is itself a tuple, then with the current design it is required to implement `PalletInfoAccess`, which of course makes no sense for anything else than a pallet.

Reply:
> The goal of this PR is to achieve exactly what you are suggesting here. Having a generic to run try-state checks at the runtime level

Then the easiest way is to just make a new trait, since the `TryState` trait is aimed at pallets (should probably have been named `PalletTryState` but now we are stuck with this for compatibility reasons).  
```rust
pub trait RuntimeTryState {
    fn try_state(whatever...) -> Result<(), TryRuntimeError>;
}
```

Anyway, i dont really want to block this approach if it has clear advantages over ^, so going to ask @kianenigma what he thinkgs.

Reply:
We've had a lot of iterations here, thank you both, esp @ntn-x2 for exploring solutions with us. 

The systemic mistake I am realizing (noted for future) is that we should have first discussed the "ideal outcome" vs the approach. So let's do that. 

what should we expect of `TryStateSelect` in the context of this PR? Should we be able to tweak `RuntimeTryState` with it? 

If yes, the current approach is the right way. If not, Oliver's approach is much simpler and `RuntimeTryState` will always be executed.

I am leaning toward not allowing runtime level try-state logic to be configurable with `TryStateSelect`. 

**First reason**: This API, as pointed out here, is already somewhat broken, and we are not even sure if it is "enough". In that, the point of it was to _regulate_ how much execution time is given to try-state hooks. But, through experience, we have seen that the issue is that a single pallet has a try-state-bomb that takes forever. In an ideal sense, instead of `TryStateSelect` you would pass some `Weight`-equivalent amount, and this much weight was used in try-state. With PVM we can do this.

**Second reason**: If we allow `TryStateSelect` to _properly_ configure runtime-try-state, the variants would be a lot. If you want "round-robin pallets + runtime", or "pallets x, y, z and runtime". These are hard to express.  

**Third Reason**: Let's do the simpler approach. IFF someone needs to disable runtime level try-state, we have a blueprint of how to solve it.

Reply:
> But, through experience, we have seen that the issue is that a single pallet has a try-state-bomb that takes forever. In an ideal sense, instead of `TryStateSelect` you would pass some `Weight`-equivalent amount, and this much weight was used in try-state. With PVM we can do this.

PVM will not come tomorrow and I also don't see why we should need this here. These `try-state` hooks are not being run as part of the on chain logic. How long they take is really not important at all. We need to ensure that they finish on time. If there is some kind of "try-state bomb", it probably means that we read too much state. Still the amount of state is not in the terabytes or similar, thus I would say that our current read performance needs to be optimized.

Reply:
@kianenigma sorry I think I did not fully understand what your suggestion here is. Are you suggesting to have a new trait which takes only a block number and that is required to be implemented by the new generic? That would mean that the `TryStateSelect` logic would only be applied to pallets, while all other runtime try-states would all always be executed. Is that the proposed way forward? Would the runtime logic still be called before the `AllPallets` logic?